### PR TITLE
fix: hide password of mongo url

### PR DESCRIPTION
**Pull request by trylovetom**
##### Checklist
- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Description of change
`/package.json`: add url mask (https://github.com/tanzim/url-mask)
`/lib/mongoose.js`: setup url mask with logger

### CODE REVIEW

**Comment by trylovetom** [comment]
@atian25 Snyk blocks this PR, I think url-mask has vulnerable dependency

File: lib/mongoose.js [hunk 1]
```diff
@@ -56,7 +57,7 @@ function createOneClient(config, app) {
   if (!options.hasOwnProperty('useNewUrlParser')) {
     options.useNewUrlParser = true;
   }
-  app.coreLogger.info('[egg-mongoose] connecting %s', url);
+  app.coreLogger.info('[egg-mongoose] connecting %s', urlMask(url, { maskUsername: false }));
```
Comment (atian25) [inline lib/mongoose.js]:
```suggestion
  const maskedUrl = urlMask(url, { maskUsername: false });
  app.coreLogger.info(`[egg-mongoose] connecting ${maskedUrl}`);
```

File: lib/mongoose.js [hunk 2]
```diff
@@ -68,16 +69,16 @@ function createOneClient(config, app) {
 
   /* istanbul ignore next */
   db.on('disconnected', () => {
-    app.coreLogger.error(`[egg-mongoose] ${url} disconnected`);
+    app.coreLogger.error(`[egg-mongoose] ${urlMask(url, { maskUsername: false })} disconnected`);
```
Comment (atian25) [inline lib/mongoose.js]:
```suggestion
    app.coreLogger.error(`[egg-mongoose] ${maskedUrl} disconnected`);
```

File: lib/mongoose.js [hunk 3]
```diff
@@ -68,16 +69,16 @@ function createOneClient(config, app) {
...
   });
 
   db.on('connected', () => {
-    app.coreLogger.info(`[egg-mongoose] ${url} connected successfully`);
+    app.coreLogger.info(`[egg-mongoose] ${urlMask(url, { maskUsername: false })} connected successfully`);
```
Comment (atian25) [inline lib/mongoose.js]:
```suggestion
    app.coreLogger.info(`[egg-mongoose] ${maskedUrl} connected successfully`);
```

File: lib/mongoose.js [hunk 4]
```diff
@@ -68,16 +69,16 @@ function createOneClient(config, app) {
...
   });
 
   /* istanbul ignore next */
   db.on('reconnected', () => {
-    app.coreLogger.info(`[egg-mongoose] ${url} reconnected successfully`);
+    app.coreLogger.info(`[egg-mongoose] ${urlMask(url, { maskUsername: false })} reconnected successfully`);
```
Comment (atian25) [inline lib/mongoose.js]:
```suggestion
    app.coreLogger.info(`[egg-mongoose] ${maskedUrl} reconnected successfully`);
```

File: package.json [hunk 5]
```diff
@@ -15,10 +15,10 @@
   "dependencies": {
     "@types/mongoose": "^5.0.18",
     "await-first": "^1.0.0",
-    "mongoose": "^5.0.17"
+    "mongoose": "^5.0.17",
```
Comment (atian25) [inline package.json]:
```suggestion
    "mongoose": "^5.4.20",
```

**Comment by atian25** [comment]
> @atian25 Snyk blocks this PR, I think url-mask has vulnerable dependency
> But I am not in the organisation of synk. I can't see any message about this.

yes, it use lodash-merge

image

maybe we do not use this lib, but write it ourself. just using `url.parse` then remove `auth` then `url.format`

File: lib/filterURLPassword.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+const { URL } = require('url');
+
+/**
+ * filter the auth of url string
+ * @param {URL} input filtered url
```
Comment (atian25) [inline lib/filterURLPassword.js]: miss @return

File: lib/filterURLPassword.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+ */
+module.exports = function filterURLPassword(input) {
+  const url = new URL(input);
```
Comment (atian25) [inline lib/filterURLPassword.js]: how about use legacy API `url.parse`.

egg-mongoose require 6.x, and URL is intro after 6.13. I'm not sure if it's break.

Comment (trylovetom) [reply lib/filterURLPassword.js]: **The status of `ur.parse` is `Stability: 0 - Deprecated`.**
https://nodejs.org/docs/latest/api/url.html#url_legacy_url_api

## Suggest
1. Check node version with `process.version` and switch function between url.parse or URL class.
2. Specify node version like `{ "engines" : { "node" : ">=6.13.0" } }`.
3. Refactor the filterURLPassword, from node API to JS API, such like Regex.

Comment (trylovetom) [reply lib/filterURLPassword.js]: @atian25 What do you think?

Comment (atian25) [reply lib/filterURLPassword.js]: node 11 is not LTS,  we will change it later.

**Comment by atian25** [comment]
3.1.2

**Comment by atian25** [comment]
@dead-horse @fengmk2 我没权限发，给个 npm 权限我下。

```bash
$ npm owner add atian25 egg-mongoose
```

**Comment by atian25** [comment]
done @trylovetom

**Comment by atian25** [comment]
@trylovetom I had rollback this version: https://github.com/eggjs/egg/issues/3675

the full mongoose connection string format is https://docs.mongodb.com/manual/reference/connection-string/

and the url parser is so complex: https://github.com/mongodb-js/mongodb-core/blob/70a7d947bc1f8568c347d882914cc335fe68ab81/lib/uri_parser.js#L487

so maybe we should use the old way - string replace.

**Pull request merged**